### DISCUSSION TITLE
CSMA performance metrics

### DISCUSSION THREAD

**Discussion by rahav** [contiki-ng/contiki-ng#2329] [Q&A] [answered]
Despite running a very small network (4 - 6 nodes + 1 root) I ran into a problem where CSMA would fail. The problem was fixed in https://github.com/contiki-ng/contiki-ng/discussions/2325 but when I was testing the change I noticed the following CSMA performance.

I wanted to surface the data to Contiki developers in case something need to be fixed, and I'll be happy to further chat on the matter. On the other hand, I would appreciate a response if this performance is as designed.

The following table shows the ratio of CSMA errors to application messages sent (e.g. there are 153.34% more status 1 errors than application messages sent when the neighbor queue size is 2).

| CSMA_CONF_MAX_NEIGHBOR_QUEUES | status 1  | status 2 | status 6 |
| ----------------------------------------- | -------- | -------- | -------- |
| 2 (default)                                                     | 153.34% | 3.83%  | 25.64% |
| 4                                                                    | 174.27%  | 4.51%  |    0.02% |

My concerns are
1. Number of error status 1 relative to actual application messages.
2. Over a quarter of the application messages aren't delivered (status 6) with the default CSMA_CONF_MAX_NEIGHBOR_QUEUES. This occurs with high probability if messages are sent close to each other in time. Is it reasonable in such very small network of 6 nodes?

**Comment by arurke** [comment, marked as answer]
@rahav Can you give a bit more info - how does the network topology look like, and how often and how large packets do each node send, and to whom? 250 kbps datarate physical layer? My initial thoughts is that you are simply meeting the limits of CSMA, but hard to tell without more info.

**Reply by rahav** [reply]
When I saw that there might be a problem I tested with what I thought would be a topology with minimal routing problems, see attached. Each square is 10 m. The data I sent is from the following topology:
<img width="391" alt="circle-topology csc" src="https://user-images.githubusercontent.com/2912950/207995941-26667e12-4b8e-4916-8daf-5ebaee10d8a3.png">

Packet size vary from 12 - 56 bytes. Sending occurs in bursts, with the most demanding being the following scenario: a node will send up to 3 messages, one immediately following the other. It will then wait for an answer from up to 3 destinations. The responses may be sent close in time to each other, but not as demanding the the initial burst of 3.

Routing is RPL Lite.

**Reply by arurke** [reply]
My gut feeling is still that you are meeting the CSMA capacity. I am no CSMA expert but my advice would be to consider the interference ranges - more nodes within each other ranges leads to more potential collisions. Also consider the routing topology and your traffic pattern - more hops to destination means more transmissions per application packet. And lastly the traffic intensities, you mention bursts, which is probably the toughest scenario for CSMA to support. If everybody are in each other interference ranges and you are sending a burst of packets back and forth between multiple nodes, and possibly everything via node 1, I would expect a fair amount of collisions.

To your concern no. 2 - depending on the topology and traffic patterns I can imagine several scenarios were 2 simultaneous queues (`CSMA_CONF_MAX_NEIGHBOR_QUEUES`) would be too few. E.g. if all traffic goes via node 1, and it received a burst of 3 packets from node 2 which is to be forwarded to nodes 4, 6, and 7 - there would not be room to queue the third packet.

PS! You can probably find a lot of good info. on CSMA performance online, the first hit on a quick scholar search for star topology gave me [this paper](https://link.springer.com/article/10.1007/s12243-016-0522-y) which actually uses cooja.
PS2! You can tweak CSMA behavior through the parameters at the top of `csma-output.c`,

**Reply by rahav** [reply]
By "interference ranges" do you mean I should position nodes in the gray zone in Cooja? I probably misunderstood the way Cooja uses the green, gray, and while zones. I thought that one should place pairwise nodes that need to communicate in the green zone relative to each other, and that the gray zone is where nodes will NOT get a message if there's a collision.

Do you know of an help/guide that explains it more formally?

**Reply by arurke** [reply]
The green zones is where the signals are sufficient such that nodes within the zone can communicate. The grey zone is the additional interference generated by a transmission. Receivers in this zone will not be able to receive the transmissions, but a transmission will interfere with any other transmissions. So you should place nodes you want to communicate in their green zones, and then be observant on your grey zones. If this is a star topology (everybody only reaching the middle node), I suspect the interference range is of less importance - indeed, thinking aloud you might benefit from large interference range as it allows the nodes to sense other nodes transmitting to the middle node and thus avoid collisions (but don't take my word for it :) )

> (...) the gray zone is where nodes will NOT get a message if there's a collision.

Yes, but nodes will also not get the message if they are in a green zone, but another grey zone overlaps and transmits at the same time. Green does not guarantee success.

> Do you know of an help/guide that explains it more formally?

There is some description [in the code](https://github.com/contiki-ng/cooja/blob/master/java/org/contikios/cooja/radiomediums/UDGM.java). You get some help if from misc. sites if you google "UDGM cooja" as well. The [tutorials ](https://docs.contiki-ng.org/en/develop/doc/tutorials/Running-a-RPL-network-in-Cooja.html) have some mention of it.

**Reply by rahav** [reply]
Thank you for everything !!!
